cluster-autoscaler/cloudprovider/gce/cache.go

```diff
@@ -0,0 +1,258 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gce
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+
+	"github.com/golang/glog"
+	gce "google.golang.org/api/compute/v1"
+)
+
+// GceCache is used for caching cluster resources state.
+type GceCache struct {
+	// Cache content.
+	migs            []*MigInformation
+	migCache        map[GceRef]Mig
+	resourceLimiter *cloudprovider.ResourceLimiter
+	machinesCache   map[MachineTypeKey]*gce.MachineType
+	// Freshness timestamps.
+	machinesCacheLastRefresh time.Time
+	// Locks. Rules of locking:
+	// - migsMutex protects only migs.
+	// - cacheMutex protects migCache, resourceLimiter and machinesCache.
+	// - methods that take cacheMutex may call methods that take migsMutex.
+	// - methods that take migsMutex may not call methods that take cacheMutex.
```

Comment:
What's the reasoning behind last 2 rules? Do you mean "if you ever need both cacheMutex and migMutex always grab cacheMutex first to avoid deadlocks"?

Reply:
Yes. Rephrased since it wasn't clear